lib/handler/connect.c

```diff
@@ -533,9 +533,11 @@ static void tcp_on_read(h2o_socket_t *_sock, const char *err)
         h2o_iovec_t vec = h2o_iovec_init(self->sock->input->bytes, self->sock->input->size);
         h2o_send(self->src_req, &vec, 1, H2O_SEND_STATE_IN_PROGRESS);
     } else {
-        /* unidirectional close is signalled using H2O_SEND_STATE_FINAL, but the write side remains open */
-        self->read_closed = 1;
-        h2o_send(self->src_req, NULL, 0, H2O_SEND_STATE_FINAL);
+        if (!self->read_closed) {
+            /* unidirectional close is signalled using H2O_SEND_STATE_FINAL, but the write side remains open */
+            self->read_closed = 1;
+            h2o_send(self->src_req, NULL, 0, H2O_SEND_STATE_FINAL);
+        }
```

Comment:
I'm not sure if we need to make this change.

There are three locations where `read_closed` is set to 1. However,
* In `close_readwrite`, `close_socket` is invoked prior to setting `read_closed` to 1. Therefore, once `read_closed` is set by `closed_readwrite`, `tcp_on_read` would never be invoked.
* In this function, `read_closed` is set only after calling `h2o_socket_read_stop`. As `h2o_send(H2O_SEND_STATE_FINAL)` is invoked immediately afterwards, `tcp_on_proceed` never be invoked thereafter and hence `h2o_socket_read_start` would never be called. Therefore, if `tcp_on_read` sets `read_closed` to 1, `tcp_on_read` would not be invoked again.
* In `tcp_on_proceed`, `read_closed` is set to 1 only when `sock` is NULL. In such case, `tcp_on_read` would never be invoked.